### PR TITLE
[PyOV] Aliases and namespace redesign

### PR DESCRIPTION
**Pull request by jiwaszki**
### Details:
 - WIP, run on CIs.
 
### Tickets:
 - *110858*

### CODE REVIEW

File: tools/mo/openvino/__init__.py [hunk 1]
```diff
@@ -1,4 +0,0 @@
-# Copyright (C) 2018-2023 Intel Corporation
-# SPDX-License-Identifier: Apache-2.0
-
-__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
Comment (rkazants) [inline tools/mo/openvino/__init__.py:4]: Why are you removing it?

File: src/bindings/python/src/openvino/__init__.py [hunk 2]
```diff
@@ -3,3 +3,47 @@
...
+from . import frontend
+from . import helpers
+from . import preprocess
+
+# Try to import openvino.tools
+try:
+    from . import tools
+    from .tools.mo import convert_model  # TODO , InputCutInfo, LayoutMap
+except ImportError:
+    warnings.warn("openvino.tools module could not be found!", ImportWarning, stacklevel=2)
```
Comment (rkazants) [inline src/bindings/python/src/openvino/__init__.py:49]: It can contradict what we are doing now by copying `convert_model` to `openvino.runtime`

**Review by rkazants — DISMISSED** [review]
Not clear changes for MO

File: src/bindings/python/src/openvino/__init__.py [hunk 3]
```diff
@@ -3,3 +3,47 @@
 # SPDX-License-Identifier: Apache-2.0
 
 __path__ = __import__("pkgutil").extend_path(__path__, __name__)  # type: ignore # mypy issue #1422
+
```
Comment (ilya-lavrenov) [inline src/bindings/python/src/openvino/__init__.py:6]: do we have other setup.py scripts which installs something in `openvino` namespace package, which actually can work without OpenVINO Python API?

Maybe accuracy checker? If yes, we cannot keep a single `openvino/__init__.py` here

File: src/bindings/python/src/openvino/__init__.py [hunk 4]
```diff
@@ -3,3 +3,47 @@
...
+# TODO should opsets go here?
+
+# Import all additional modules
+from . import frontend
+from . import helpers
+from . import preprocess
+
+# Try to import openvino.tools
+try:
+    from . import tools
```
Comment (ilya-lavrenov) [inline src/bindings/python/src/openvino/__init__.py:46]: what if tools are installed into a different python search folder?

File: src/bindings/python/src/openvino/__init__.py (hunk 2 above)
Comment (jiwaszki) [reply src/bindings/python/src/openvino/__init__.py:49]: @slyalin please remind your request here, thanks!

File: tools/mo/openvino/__init__.py (hunk 1 above)
Comment (rkazants) [reply tools/mo/openvino/__init__.py:4]: please put details into description of PR. It is also a good practice because you are implementer and can create short reasonable summary. In JIRA we can have a mess with comments, inaccurate description, etc.

File: src/bindings/python/src/openvino/__init__.py (hunk 3 above)
Comment (jiwaszki) [reply src/bindings/python/src/openvino/__init__.py:6]: This is a good question. I also modified some packages on OMZ side (for now locally) and it works around the problem for `mo`. Need to double-check the rest.

Generally you refer to packaging styles. If we consider https://peps.python.org/pep-0420/#packaging-implications there are two options:

1) `__init__` files are involving same contents and it works as overriding them is always resulting in the same file.
2) There are no `__init__` files at all on namespace level.

Second option is what is implemented here. The promise is that everything will reside in same namespace `openvino` with only main repository being holding `__init__.py` file. This is currently working with `mo` for example.

I need to test it, this is why I put up this PR to have some runs on CIs and catch all issues.
BTW, this is the cost for requested change of namespace design. At least I do not know if there is another way around it.

File: src/bindings/python/src/openvino/__init__.py (hunk 4 above)
Comment (jiwaszki) [reply src/bindings/python/src/openvino/__init__.py:46]: What exact scenario do you have in mind here? For example, what if runtime is a wheel and tools are in PYTHONPATH?

Comment (ilya-lavrenov) [reply src/bindings/python/src/openvino/__init__.py:46]: > What exact scenario do you have in mind here? For example, what if runtime is a wheel and tools are in PYTHONPATH?

Yes, something like this.
Or runtime is wheel, while tools are current directory (also searched by `site` module).

File: src/bindings/python/src/openvino/__init__.py (hunk 3 above)
Comment (eaidova) [reply src/bindings/python/src/openvino/__init__.py:6]: >>do we have other setup.py scripts which install something in openvino namespace package, which actually can work without OpenVINO Python API?
Maybe accuracy checker? 

yes, please take into account that accuracy checker can work without openvino installation and installed in openvino.tools

File: src/bindings/python/src/openvino/__init__.py (hunk 2 above)
Comment (slyalin) [reply src/bindings/python/src/openvino/__init__.py:49]: convert_model will have an origin in openvino.tools.ovc and is exposed here. Currently we are exposing it from tools.mo then we will switch to ovc. It looks irrelevant to this PR and doesn't contradict.

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 5]
```diff
@@ -2,4 +2,53 @@
 # Copyright (C) 2018-2023 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
 
-__path__ = __import__("pkgutil").extend_path(__path__, __name__)  # type: ignore # mypy issue #1422
```
Comment (popovaan) [inline src/bindings/python/src/compatibility/openvino/__init__.py:5]: Please try to return this line and add it to other root init files.
Locally it fixes problem with usage MO/OVC directly from a path.

File: tools/benchmark_tool/openvino/tools/__init__.py [hunk 6]
```diff
@@ -1,3 +1,8 @@
 # Copyright (C) 2018-2023 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
-__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
Comment (popovaan) [inline tools/benchmark_tool/openvino/tools/__init__.py:3]: Please return this line.

File: tools/mo/openvino/tools/__init__.py [hunk 7]
```diff
@@ -1,4 +1,8 @@
 # Copyright (C) 2018-2023 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
 
-__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
Comment (popovaan) [inline tools/mo/openvino/tools/__init__.py:4]: Please return this line.

File: tools/ovc/openvino/tools/__init__.py [hunk 8]
```diff
@@ -1,4 +1,8 @@
 # Copyright (C) 2018-2023 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
 
-__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
Comment (popovaan) [inline tools/ovc/openvino/tools/__init__.py:4]: Please return this line.

File: src/bindings/python/src/openvino/__init__.py [hunk 9]
```diff
@@ -2,4 +2,55 @@
...
+    from openvino.runtime import Layout
+    from openvino.runtime import Tensor
+
+    from openvino.runtime import get_batch
+    from openvino.runtime import set_batch
+    from openvino.runtime import serialize
+    from openvino.runtime import shutdown
+    from openvino.runtime import tensor_from_file
+    from openvino.runtime import compile_model
+    from openvino.runtime import get_version
```
Comment (slyalin) [inline src/bindings/python/src/openvino/__init__.py:41]: Also, please add `save_model`.

File: tools/benchmark_tool/openvino/tools/__init__.py [hunk 10]
```diff
@@ -1,3 +1,10 @@
...
+
+__path__ = __import__("pkgutil").extend_path(__path__, __name__)
+
+try:
+    import openvino.tools.mo
```
Comment (ilya-lavrenov) [inline tools/benchmark_tool/openvino/tools/__init__.py:7]: what's about POT and other tools? should we try to import it as well? if no - why?

File: tools/ovc/CMakeLists.txt [hunk 11]
```diff
@@ -35,6 +35,13 @@ install(DIRECTORY ${OpenVINOConverter_SOURCE_DIR}/openvino
         ${OV_CPACK_COMP_OVC_EXCLUDE_ALL}
         USE_SOURCE_PERMISSIONS)
 
+if(ENABLE_TESTS)
+    install(DIRECTORY unit_tests
+            DESTINATION tests/ovc
+            COMPONENT tests
+            EXCLUDE_FROM_ALL)
+endif()
```
Comment (ilya-lavrenov) [inline tools/ovc/CMakeLists.txt:43]: why is it needed in current PR? seems unrelated to namespaces work

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 12]
```diff
@@ -2,4 +2,55 @@
...
+    import warnings
+    warnings.warn("openvino package has problems with imports!", ImportWarning, stacklevel=2)
+
+# Import openvino.tools
+# Capture it in try-except with pass so circular imports are allowed.
+# TODO: restructure packages to remove WA
+try:
+    from openvino import tools as tools
+    # Model Conversion API - ovc should reside in the main namespace
+    from openvino.tools.ovc import convert_model, InputCutInfo, LayoutMap
```
Comment (popovaan) [inline src/bindings/python/src/compatibility/openvino/__init__.py:54]: Please remove `LayoutMap`. It was removed from `ovc`.

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 13]
```diff
@@ -2,4 +2,55 @@
...
-__path__ = __import__("pkgutil").extend_path(__path__, __name__)  # type: ignore # mypy issue #1422
+__path__ = __import__("pkgutil").extend_path(__path__, __name__)
+
+try:
+    # Import all modules
+    from openvino import runtime as runtime
+    from openvino import frontend as frontend
+    from openvino import helpers as helpers
+    from openvino import preprocess as preprocess
+    from openvino import test_utils as test_utils
```
[the remark this answers is not in the document]
Comment (jiwaszki) [reply src/bindings/python/src/compatibility/openvino/__init__.py]: Moved to separate `try`

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 14]
```diff
@@ -2,4 +2,68 @@
...
+    # Set version for openvino package
+    from openvino.runtime import get_version
+    __version__ = get_version()
+except ImportError:
+    import warnings
+    warnings.warn("openvino package has problems with imports!", ImportWarning, stacklevel=2)
+
+# Import openvino.tools
+# Capture it in try-except with pass so circular imports are allowed.
+# TODO: restructure packages to remove WA
```
Comment (slyalin) [inline src/bindings/python/src/compatibility/openvino/__init__.py:63]: Is it correct understanding that try-except here is needed only to solve circular imports? Why do we have that circular imports and how to avoid it? @popovaan, could you also comment?

Comment (ilya-lavrenov) [reply src/bindings/python/src/compatibility/openvino/__init__.py:63]: looks like OVC internally import OV runtime and I expect infinite recursive here rather than circular dependency

File: tools/mo/openvino/tools/__init__.py [hunk 15]
```diff
@@ -1,4 +1,10 @@
...
+__path__ = __import__("pkgutil").extend_path(__path__, __name__)
+
+try:
+    import openvino.tools.mo
```
Comment (ilya-lavrenov) [inline tools/mo/openvino/tools/__init__.py:7]: why do we need it here?

File: src/bindings/python/src/compatibility/openvino/__init__.py (hunk 14 above)
Comment (jiwaszki) [reply src/bindings/python/src/compatibility/openvino/__init__.py:67]: It was added in earlier stages and it fixed the problem. In current solution (as Ilya says) it creates infinite recursion as you are able to go `openvino.tools.openvino.tools.openvino...`.

However try-except here will not change much as it will stay here or to move one "try-except" above. OVC is part of `openvino` itself so it will be either imported correctly or passed around. Example would be using legacy MO from PYTHONPATH. So all inits must match and go around OVC as it won't be discovered and result in error in such case.

File: tools/ovc/CMakeLists.txt (hunk 11 above)
Comment (jiwaszki) [reply tools/ovc/CMakeLists.txt:43]: It is a leftover from previous tests where OVC tests were copied to avoid being run from same directory. Similar tactic to MO tests.

Would you like to remove it and test it over with current approach?

File: tools/benchmark_tool/openvino/tools/__init__.py (hunk 10 above)
Comment (jiwaszki) [reply tools/benchmark_tool/openvino/tools/__init__.py:7]: Please check messages about the topic. I called for all requests, awaiting responses.

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 16]
```diff
@@ -2,4 +2,72 @@
...
+    from openvino.runtime import InferRequest
+    from openvino.runtime import AsyncInferQueue
+
+    from openvino.runtime.op import Constant
+    from openvino.runtime.op import Parameter
+
+    from openvino.runtime import Extension
+    from openvino.runtime import Dimension
+    from openvino.runtime import Strides
+    from openvino.runtime import PartialShape
```
Comment (AlexKoff88) [inline src/bindings/python/src/compatibility/openvino/__init__.py:43]: how about `layout_helpers`. I found it is used quite often.

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 17]
```diff
@@ -2,4 +2,72 @@
...
+    from openvino.runtime import Shape
+    from openvino.runtime import Layout
```
Comment (AlexKoff88) [inline src/bindings/python/src/compatibility/openvino/__init__.py:45]: How about `custom_preprocess_function`. I found it is used quite often.

File: src/bindings/python/src/compatibility/openvino/__init__.py (hunk 16 above)
Comment (jiwaszki) [reply src/bindings/python/src/compatibility/openvino/__init__.py:43]: Added

File: tools/ovc/CMakeLists.txt (hunk 11 above)
Comment (jiwaszki) [reply tools/ovc/CMakeLists.txt]: Reverted and works.

File: src/bindings/python/src/openvino/__init__.py [hunk 18]
```diff
@@ -2,4 +2,71 @@
...
+    from openvino.runtime import tensor_from_file
+    from openvino.runtime import save_model
+    from openvino.runtime import layout_helpers
+
+    # Set version for openvino package
+    from openvino.runtime import get_version
+    __version__ = get_version()
+except ImportError:
+    import warnings
+    warnings.warn("openvino package has problems with imports!", ImportWarning, stacklevel=2)
```
Comment (slyalin) [inline src/bindings/python/src/openvino/__init__.py:64]: This is not printed by default, so when something goes wrong with one of the imports above, there is no way to immediately understand what's going wrong (my case -- I have issues with the build from this branch). Why cannot we raise in case of any issue -- why do we need try-except?

File: tools/benchmark_tool/openvino/tools/__init__.py (hunk 10 above)
Comment (ilya-lavrenov) [reply tools/benchmark_tool/openvino/tools/__init__.py:7]: I don't understand what I should check
I've asked 2 questions and have not received the answers

File: src/bindings/python/src/openvino/__init__.py (hunk 18 above)
Comment (jiwaszki) [reply src/bindings/python/src/openvino/__init__.py:64]: @slyalin if you run OVC or MO from PATH there were issues with imports according to import hierarchy as `openvino.runtime` does not exists there. However it was few days ago.

Is there a test for such use-case? If yes, I can remove try-except and test it out.

File: tools/mo/openvino/__init__.py [hunk 19]
```diff
@@ -1,4 +1,72 @@
...
+    # Import all public modules
+    from openvino import runtime as runtime
+    from openvino import frontend as frontend
+    from openvino import helpers as helpers
+    from openvino import preprocess as preprocess
+    from openvino import utils as utils
+    from openvino.runtime import properties as properties
+    # Import old API
+    # TODO: remove in 2024.0
+    from openvino import inference_engine as inference_engine
```
Comment (ilya-lavrenov) [inline tools/mo/openvino/__init__.py:23]: I suggest also moving this import to a dedicated try / catch, because user can miss installation of API 1.0 requirements and it will not be build. 
So, for each group of functionality, let's have a dedicated block

File: src/bindings/python/src/openvino/__init__.py (hunk 18 above)
Comment (ilya-lavrenov) [reply src/bindings/python/src/openvino/__init__.py:64]: Agree. I also suppose that `openvino.runtime` should always be available as a "base" module

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 20]
```diff
@@ -2,4 +2,71 @@
...
+    from openvino import utils as utils
+    from openvino.runtime import properties as properties
+    # Import old API
+    # TODO: remove in 2024.0
+    from openvino import inference_engine as inference_engine
+
+    # Required for Windows OS platforms
+    from openvino.utils import _add_openvino_libs_to_search_path
+
+    _add_openvino_libs_to_search_path()
```
Comment (ilya-lavrenov) [inline src/bindings/python/src/compatibility/openvino/__init__.py:28]: should this statement be even before `from openvino import test_utils as test_utils ` ?
E.g. we need to set libraries path first, then try to import functionality. 

I suppose if `test_utils` depends on some DLLs, it can fail without `_add_openvino_libs_to_search_path `, can not it?

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 21]
```diff
@@ -2,4 +2,71 @@
...
+    # Set version for openvino package
+    from openvino.runtime import get_version
+    __version__ = get_version()
+except ImportError:
+    import warnings
+    warnings.warn("openvino package has problems with imports!", ImportWarning, stacklevel=2)
+
+# Import openvino.tools
+try:
+    from openvino import tools as tools
```
Comment (ilya-lavrenov) [inline src/bindings/python/src/compatibility/openvino/__init__.py:68]: could you please explain why do we need this line? Just to use:
```python
import openvino
# and later use?
openvin.tools.XXX 
```
If yes, what `openvino.tools` do we want to exppose this way?

File: tools/pot/openvino/tools/__init__.py [hunk 22]
```diff
@@ -1,3 +1,10 @@
...
 # SPDX-License-Identifier: Apache-2.0
-__path__ = __import__('pkgutil').extend_path(__path__, __name__)
+
+__path__ = __import__("pkgutil").extend_path(__path__, __name__)
+
+try:
+    import openvino.tools.mo
+except ImportError:
+    import warnings
+    warnings.warn("openvino.tools.mo module could not be found!", ImportWarning, stacklevel=2)
```
Comment (ilya-lavrenov) [inline tools/pot/openvino/tools/__init__.py:10]: do we need to change this file at all? 
I suppose a default file with `pkgutil` is enough, is not it?

File: tools/mo/openvino/__init__.py (hunk 19 above)
Comment (jiwaszki) [reply tools/mo/openvino/__init__.py:23]: Agree, will push with next commit

File: src/bindings/python/src/compatibility/openvino/__init__.py (hunk 20 above)
Comment (jiwaszki) [reply src/bindings/python/src/compatibility/openvino/__init__.py:28]: Good point.

File: src/bindings/python/src/compatibility/openvino/__init__.py [hunk 23]
```diff
@@ -2,4 +2,61 @@
...
+
+# API 2.0
+try:
+    # Import all public modules
+    from openvino import runtime as runtime
+    from openvino import frontend as frontend
+    from openvino import helpers as helpers
+    from openvino import preprocess as preprocess
+    from openvino import utils as utils
+    from openvino.runtime import properties as properties
```
Comment (slyalin) [inline src/bindings/python/src/compatibility/openvino/__init__.py:23]: Experiment showed that there is no need to have these imports: tried to remove this block and imported from runtime, frontend, all worked fine. I suppose we should have a test that checks the imports work and we are done.

Comment (ilya-lavrenov) [reply src/bindings/python/src/compatibility/openvino/__init__.py:23]: Yes, it works. 

I suppose here author wants to achieve fast access without boilerplate imports on top of file

Comment (jiwaszki) [reply src/bindings/python/src/compatibility/openvino/__init__.py:23]: It works but as @ilya-lavrenov mention, my goal is to provide fast interface like this (if anyone would like to use it):
```python
import openvino as ov
ov.runtime.Something(...)
```
instead of:
```python
import openvino as ov
# some code...
import openvino.runtime as ov_runtime
ov_runtime.Something(...)
```

File: src/bindings/python/src/openvino/__init__.py (hunk 18 above)
Comment (ilya-lavrenov) [reply src/bindings/python/src/openvino/__init__.py:55]: agree to experiment with removal of try / except in later PRs.

File: src/bindings/python/src/compatibility/openvino/__init__.py (hunk 23 above)
Comment (slyalin) [reply src/bindings/python/src/compatibility/openvino/__init__.py:23]: Just tried
```
import openvino as ov
ov.frontend.pytorch.Decoder
```
and it worked even without those imports. Tried stuff from runtime as well. Don't know how it work though.

Comment (ilya-lavrenov) [reply src/bindings/python/src/compatibility/openvino/__init__.py:23]: > Don't know how it works though

I also don't know, because `openvino/frontend/__init__.py` does not import `pytorch`
@jiwaszki any idea?

**Review by slyalin — APPROVED** [review]
Tested manually in clean environment with `openvino` wheel installed alone and in combination with `openvino-dev` wheel. Various combinations were tried with `convert_model` imported from both `ovc` and `mo`, in a way how `mo` is imported in the notebooks, `compile_model`, torch models and onnx models and random stuff from various openvino component. Not exhaustive but quite representative set of tests.
